moonraker/plugins/history.py

```diff
@@ -0,0 +1,253 @@
+        self.job_id += 1
+        self.database.insert_item(HIST_NAMESPACE,"job_auto_inc_id", self.job_id)
+
+        self.current_job = job_id
+        self.jobs[job_id] = job
+        self.grab_job_metadata()
+        self.save_current_job()
+
+    def delete_job(self, id):
+        if id not in self.jobs:
+            return False
+        del self.jobs[id]
+        self.database.delete_item(HIST_NAMESPACE, "jobs.%s" % id)
+        return id
+
+    def finish_job(self, status, updates):
+        if self.current_job == None:
+            return
+
+        self.jobs[self.current_job].finish("completed", self.print_stats)
+        self.save_current_job()
+        self.current_job = None
+
+    def get_job(self, id):
+        if id not in self.jobs:
+            return None
+        return self.jobs.get(id)
+
+    def grab_job_metadata(self):
+        if self.current_job == None or self.current_job not in self.jobs:
+            return
+
+        gcdb = self.database.wrap_namespace("gcode_metadata",  parse_keys=False)
+        logging.debug("gcdb: %s" % gcdb)
+        filename = self.jobs[self.current_job].get("filename")
+        if filename not in gcdb:
+            return
+
+        metadata = {}
+        for i in METADATA_KEYS:
+            if i in gcdb[filename]:
+                self.jobs[self.current_job].update_file_metadata(
+                    {i: gcdb[filename][i]})
+
+    def save_current_job(self):
+        self.database.insert_item(HIST_NAMESPACE, "prints.%s" %
+            self.current_job, self.jobs[self.current_job].get_stats())
+
+    def save_to_database(self):
+        self.database.insert_item(HIST_NAMESPACE, "prints",
+            json.loads(json.dumps(self.jobs,default=lambda o: o.__dict__)))
+
+class PrinterJob:
+    def __init__(self, data={}, file_metadata={}):
+        self.end_time = None
+        self.filament_used = 0
+        self.filename = None
+        self.print_duration = 0
+        self.status = "in_progress"
+        self.start_time = time.time()
+        self.total_duration = 0
+        self.file_metadata = file_metadata
+        self.update_from_ps(data)
+        self.update_file_metadata(file_metadata)
+
+    def finish(self, status, print_stats={}):
+        self.end_time = time.time()
+        self.status = status
+        self.update_from_ps(print_stats)
+
+    def get(self, name):
+        if not hasattr(self, name):
+            return None
+        return getattr(self, name)
+
+    def get_stats(self):
+        return {k: getattr(self, k) for k in self.__dict__}
+
+    def get_metadata(self):
+        return self.file_metadata
+
+    def set(self, name, val):
+        if not hasattr(self, name):
+            return
+        setattr(self, name, val)
+
+    def update_from_ps(self, data):
+        for i in data:
+            if hasattr(self, i):
+                setattr(self, i, data[i])
+
+    def update_file_metadata(self, file_metadata={}):
+        for i in file_metadata.keys():
```

Comment:
I believe the estimated values are still pertinent to history. Personally, I would like to know as I would like to see the delta between estimated and actual in order to know how accurate the slicer is and be able to plan appropriately. Granted, this will vary depending on each model, but it would be something I would like to know.